### PR TITLE
Add Sorbet types to updater types

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
### What are you trying to accomplish?

Reduce opportunities for type-errors and nil-references by using static and runtime sorbet tools.

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

**Review by abdulapopoola — APPROVED** [review]

**Comment by ryanbrandenburg** [comment]
> @JamieMagee does this mean we can add `strictSigil: true` or something like that to the rubocop file to lock this in?

Unfortunately I'm missing environment.rb in this PR. I struggled to get the results of ENV to give me typed values without abusing T.cast or something similar.

**Comment by JamieMagee** [comment]
@ryanbrandenburg I almost got environment in #8318. Maybe you can finish the work I started?

File: updater/lib/dependabot/updater/group_update_refreshing.rb
Comment (JamieMagee) [inline updater/lib/dependabot/updater/group_update_refreshing.rb:1]: Why did this get deleted?

File: updater/lib/dependabot/base_command.rb [hunk 1]
```diff
@@ -8,21 +8,28 @@
...
 module Dependabot
   class RunFailure < StandardError; end
 
   class BaseCommand
+    extend T::Sig
+
     # Implement in subclass
+    sig { overridable.void }
     def perform_job
       raise NotImplementedError
```
Comment (JamieMagee) [inline updater/lib/dependabot/base_command.rb:22]: Can you remove the implementation here? Same for all the methods below that just `raise NotImplementedError`.

File: updater/lib/dependabot/base_command.rb [hunk 2]
```diff
@@ -47,21 +55,24 @@ def run
...
+    sig { params(err: StandardError).void }
     def handle_unknown_error(err)
       error_details = {
         ErrorAttributes::CLASS => err.class.to_s,
         ErrorAttributes::MESSAGE => err.message,
-        ErrorAttributes::BACKTRACE => err.backtrace.join("\n"),
-        ErrorAttributes::FINGERPRINT => err.respond_to?(:sentry_context) ? err.sentry_context[:fingerprint] : nil,
+        ErrorAttributes::BACKTRACE => err.backtrace&.join("\n"),
+        ErrorAttributes::FINGERPRINT => err.respond_to?(:sentry_context) ?
+          T.unsafe(err).sentry_context[:fingerprint] : nil,
```
Comment (JamieMagee) [inline updater/lib/dependabot/base_command.rb:75]: nit

```suggestion
        ErrorAttributes::FINGERPRINT => err.respond_to?(:sentry_context)
          ? T.unsafe(err).sentry_context[:fingerprint]
          : nil,
```

File: updater/lib/dependabot/file_fetcher_command.rb [hunk 3]
```diff
@@ -141,65 +157,92 @@ def dependency_files_for_globs
...
+        yield
+      rescue Octokit::BadGateway
+        retries += 1
+        retry if retries <= 2
+        raise
+      end
+    end
+
+    sig { params(blk: T.proc.returns(T::Array[DependencyFile])).returns(T::Array[DependencyFile]) }
+    def with_retries(&blk)
```
Comment (JamieMagee) [inline updater/lib/dependabot/file_fetcher_command.rb:194]: Would making this generic prevent the duplication?

File: updater/lib/dependabot/file_fetcher_command.rb [hunk 4]
```diff
@@ -233,18 +276,22 @@ def handle_file_fetcher_error(error)
       service.capture_exception(error: error, job: job)
     end
 
+    sig { params(error: StandardError).returns(Float) }
     def rate_limit_error_remaining(error)
+      client_error = T.cast(error, Octokit::ClientError)
```
Comment (JamieMagee) [inline updater/lib/dependabot/file_fetcher_command.rb:281]: I would be more specific about the parameter type, if possible, and do the cast in the caller.

File: updater/lib/dependabot/updater/group_update_refreshing.rb
Comment (ryanbrandenburg) [reply updater/lib/dependabot/updater/group_update_refreshing.rb:1]: Its methods were only used by a single class (and it seems unlikely to expand), so I just included them in that class.

File: updater/lib/dependabot/updater.rb [hunk 5]
```diff
@@ -16,41 +16,47 @@
...
 
       Dependabot.logger.debug("Performing job with #{operation_class}")
       service.increment_metric("updater.started", tags: { operation: operation_class.tag_name })
-      operation_class.new(
+      operation = operation_class.new(
         service: service,
         job: job,
         dependency_snapshot: dependency_snapshot,
         error_handler: error_handler
-      ).perform
```
Comment (JamieMagee) [inline updater/lib/dependabot/updater.rb:45]: Why was this change required?

File: updater/lib/dependabot/updater.rb [hunk 6]
```diff
@@ -16,41 +16,47 @@
...
+      operation.perform
     rescue *ErrorHandler::RUN_HALTING_ERRORS.keys => e
       # TODO: Drop this into Security-specific operations
       if e.is_a?(Dependabot::AllVersionsIgnored) && !job.security_updates_only?
         error = StandardError.new(
           "Dependabot::AllVersionsIgnored was unexpectedly raised for a non-security update job"
         )
         error.set_backtrace(e.backtrace)
-        service.capture_exception(error: error, job: job)
+        @service.capture_exception(error: error, job: job)
```
Comment (JamieMagee) [inline updater/lib/dependabot/updater.rb:59]: Nice catch! Was this a bug that Sorbet found?

File: updater/lib/dependabot/updater.rb [hunk 7]
```diff
@@ -63,9 +69,13 @@ def run
...
     private
 
+    sig { returns(Service) }
     attr_reader :service
+    sig { returns(Job) }
     attr_reader :job
+    sig { returns(DependencySnapshot) }
     attr_reader :dependency_snapshot
+    sig { returns(ErrorHandler) }
     attr_reader :error_handler
```
Comment (JamieMagee) [inline updater/lib/dependabot/updater.rb:79]: nit:

```suggestion
    sig { returns(Service) }
    attr_reader :service

    sig { returns(Job) }
    attr_reader :job

    sig { returns(DependencySnapshot) }
    attr_reader :dependency_snapshot

    sig { returns(ErrorHandler) }
    attr_reader :error_handler
```

File: updater/lib/dependabot/updater/error_handler.rb [hunk 8]
```diff
@@ -66,20 +77,36 @@ def handle_dependency_error(error:, dependency:, dependency_group: nil)
       end
 
       # Provides logging for errors that occur when processing a dependency
+      sig do
+        params(
+          dependency: T.nilable(Dependabot::Dependency),
+          error: StandardError,
+          error_type: String,
+          error_detail: T.nilable(T.any(String, T::Hash[T.untyped, T.untyped]))
```
Comment (JamieMagee) [inline updater/lib/dependabot/updater/error_handler.rb:85]: Can you be more specific with the hash?

File: updater/lib/dependabot/updater/error_handler.rb [hunk 9]
```diff
@@ -118,7 +146,9 @@ def log_job_error(error:, error_type:, error_detail: nil)
 
       private
 
+      sig { returns(Service) }
       attr_reader :service
+      sig { returns(Job) }
       attr_reader :job
```
Comment (JamieMagee) [inline updater/lib/dependabot/updater/error_handler.rb:152]: nit:

```suggestion
      sig { returns(Service) }
      attr_reader :service

      sig { returns(Job) }
      attr_reader :job
```

File: updater/lib/dependabot/updater/group_update_creation.rb [hunk 10]
```diff
@@ -19,24 +19,12 @@ module Dependabot
   class Updater
     extend T::Sig
 
-    module GroupUpdateCreation
+    class GroupUpdateCreation < OperationBase
```
Comment (JamieMagee) [inline updater/lib/dependabot/updater/group_update_creation.rb:22]: I am not sure this is the best approach. This is a mixin in Ruby. I had originally attempted to type this in #8319, but now that I have more experience with Ruby, I realise this isn't the correct approach. Check out #9917 for an example of how I typed a helper mixin in `pub`.

Comment (JamieMagee) [reply updater/lib/dependabot/updater/group_update_creation.rb:22]: Actually, reviewing further down this PR, it looks like you've just added this as part of the class hierarchy. I believe that will work as well!

Comment (ryanbrandenburg) [reply updater/lib/dependabot/updater/group_update_creation.rb:22]: That approach gets us to `#typed: true`, but once you do `#typed: strict` Sorbet complains that you never populated those attr_reader's.

File: updater/lib/dependabot/file_fetcher_command.rb [hunk 11]
```diff
@@ -141,76 +158,91 @@ def dependency_files_for_globs
...
     end
 
+    sig { params(file_fetcher: FileFetchers::Base).void }
     def post_ecosystem_versions(file_fetcher)
       ecosystem_versions = file_fetcher.ecosystem_versions
       api_client.record_ecosystem_versions(ecosystem_versions) unless ecosystem_versions.nil?
     end
 
-    def with_retries(max_retries: 2)
+    sig { params(_blk: T.proc.returns(T.untyped)).returns(T.untyped) }
```
Comment (JamieMagee) [inline updater/lib/dependabot/file_fetcher_command.rb:182]: Not quite what I meant. check out [the Sorbet documentation on generics][1]. I think it might be something like

```suggestion
    sig do
      type_parameters(:T)
      	.params(_blk: T.proc.returns(T.type_parameter(:T)))
      	.returns(T.type_parameter(:T)) }
    end
```

[1]: https://sorbet.org/docs/generics

File: updater/lib/dependabot/updater/error_handler.rb (hunk 8 above)
Comment (ryanbrandenburg) [reply updater/lib/dependabot/updater/error_handler.rb:85]: Nothing obvious breaks if I change it to `T::Hash[Symbol, T.untyped]`, [but other loggers](https://github.com/dependabot/dependabot-core/blob/7f7108ceb8edde9ddc8f5e0f9f67d0dc2e7eb9cb/updater/lib/dependabot/api_client.rb#L102) that take an `error_details` in this area take `T::Hash[T.untyped, T.untyped]` so I figured it best to leave it open rather than cause a type-error on a log method since we're just going to be doing `to_s` anyway.

I tried not to spread `T.untyped` around too much, but in this case it really seems that we don't case about what the type is, so it seemed safe.

**Comment by ryanbrandenburg** [comment]
Closing this in favor of smaller changes.
